feature/text_based.py

```diff
@@ -0,0 +1,724 @@
+
+
+    def _get_diversity_cost(self, selected_size: int) -> List[float]:
+        """
+        diversity cost:
+            there is a situation where selected_size is equal to the number of rows in features and all diversity
+             costs become zero. There are two ways to handle this situation:
+                * add dummy cost to diversity cost (DUMMY)
+                * check if any of the distances in the distances matrix is zero, and if so, replace it with
+                 the next smallest non-zero distance (NEXT_S_DIS).
+        """
+        kmeans = KMeans(n_clusters=selected_size, random_state=self.seed, n_init=self.trials)
+        distances = kmeans.fit_transform(self.features)
+
+        ##########NEXT_S_DIS#####
+        for i in range(self._num_cols):
+            if np.min(distances[i]) == 0:
+                nonzero_distances = distances[i][np.nonzero(distances[i])]
+                if len(nonzero_distances) > 0:
+                    min_nonzero_distance = np.min(nonzero_distances)
+                    distances[i][distances[i] == 0] = min_nonzero_distance
+        ##########NEXT_S_DIS#####
+
+        diversity_cost = [np.min(distances[i]) for i in range(self._num_cols)]
+
+        ##########DUMMY##########
+        # diversity_cost = [c + 1e-1 for c in diversity_cost]
+        ##########DUMMY##########
+
+        # Scale contexts so that sum of costs remain constant
+        diversity_cost = [c * self._num_cols / sum(diversity_cost) for c in diversity_cost]
+
+        return diversity_cost
+
+    def _get_num_row_covered(self, selected) -> int:
+        assert self.matrix is not None, "matrix is not initialized"
+
+        # Create indicator to count number of covered rows
+        row_covered = np.sum(self.matrix[:, selected], axis=1)
+
+        row_covered[row_covered > 0] = 1
+
+        num_row_covered = sum(row_covered)
+
+        return num_row_covered
+
+    @staticmethod
+    def _validate_args(selection_size):
+        if selection_size is not None:
+            if selection_size <= 0:
+                raise ValueError("Selection size must be greater than zero.")
+
+
+# TODO not sure if this should use _BaseDispatcher or not. We can decide later
+class _TextBased(_BaseSupervisedSelector):
+    """
+    featrization and optimization are done in 'fit' function.
+    The 'transform' function is used to show the selected contents
+    """
+
+    def __init__(self, seed: int, num_features: Num,
+                 featurization_method: TextWiser, optimization_method: str, cost_metric: Optional[str] = None):
+        # Call constructor of parent class _BaseSupervisedSelector
+        super().__init__(seed)
+
+        self.num_features = num_features
+        self.featurization_method = featurization_method
+        self.optimization_method = optimization_method
+        self.cost_metric = cost_metric
+        self.content_selector = ContentSelector(selection_size=num_features, seed=123456, trials=100, verbose=True)
```

Comment:
you are missing to use the self.seed here

Reply:
after all, seed is essentially the most important part of the Random selector :D

Reply:
Fixed! Thank you for mentioning it!